lib/handler/RetryHandler.js

```diff
@@ -0,0 +1,347 @@
+        : Math.min(currentTimeout * timeoutFactor ** counter, maxTimeout)
+
+    return retryTimeout
+  }
+
+  onHeaders (statusCode, rawHeaders, resume, statusMessage) {
+    const headers = parseHeaders(rawHeaders)
+    const err = new RequestRetryError(
+      'Request failed after retry',
+      statusCode,
+      {
+        headers,
+        count: this.count
+      }
+    )
+
+    this.count += 1
+
+    if (statusCode > 299) {
+      this.abort(err)
+      return false
+    }
+
+    // Checkpoint for resume from where we left it
+    if (this.resume != null) {
+      this.resume = null
+
+      if (statusCode !== 206) {
+        return true
+      }
+
+      const contentRange = parseRangeHeader(headers['content-range'])
+      if (
+        // Let's start with a weak etag check
+        (this.etag != null && this.etag !== headers.etag) ||
+        // If no content range
+        !contentRange
+      ) {
+        this.abort(err)
+        return false
+      }
+
+      const { start, size, end = size } = contentRange
+
+      assert(this.head === start, 'content-range mismatch')
+      assert(this.tail == null || this.tail === end, 'content-range mismatch')
+
+      this.resume = resume
+      return true
+    }
+
+    if (this.tail == null) {
+      if (statusCode === 206) {
+        // First time we receive 206
+        const range = parseRangeHeader(headers['content-range'])
+
+        if (range == null) {
+          return this.handler.onHeaders(
+            statusCode,
+            rawHeaders,
+            resume,
+            statusMessage
+          )
+        }
+
+        const { start, size, end = size } = range
+
+        assert(
+          start != null && Number.isFinite(start) && this.head !== start,
+          'content-range mismatch'
+        )
+        assert(Number.isFinite(start))
+        assert(
+          end != null && Number.isFinite(end) && this.tail !== end,
+          'invalid content-length'
+        )
+
+        this.head = start
+        this.tail = end
+      }
+
+      // We make our best to checkpoint the body for further range headers
+      if (this.tail == null) {
+        const contentLength = headers['content-length']
+        this.tail = contentLength != null ? Number(contentLength) : null
+      }
+
+      assert(Number.isFinite(this.head))
+      assert(
+        this.tail == null || Number.isFinite(this.tail),
+        'invalid content-length'
+      )
+
+      this.resume = resume
+      this.etag = headers.etag != null ? headers.etag : null
+
+      return this.handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage
+      )
+    }
+
+    // Meant to just indicate weather or not trigger a retry
+    // It should return null to indicate we exhausted the retries
+    const retryAfter = this.retryOpts.retry(
+      err,
+      { counter: this.count, currentTimeout: this.retryAfter },
+      { retry: this.retryOpts, ...this.opts }
+    )
```

Comment:
I would like the retry method to be asynchronous and the timeout lives inside of it, i.e. instead of returning a duration.

Reply:
Either by returning a promise or taking a callback

Reply:
Hmm, has sense; it can enable several use cases

I'm imagine something like
```suggestion
    const retryAfter = this.retryOpts.retry(
      err,
      { counter: this.count, currentTimeout: this.retryAfter },
      { retry: this.retryOpts, ...this.opts }
    ).then(timeout => { /* timeout goes here */}, err => this.handler.onError(er))
```

Thoughts?

Reply:
ACtually you don't need it to return any value, when the promise finishes then that's the retry.

Reply:
Yeah, I was actually about to ask you the same; basically the promise handles the retry; after it resolves we call `this.dispatch`, otherwise we handle the error

Reply:
TBH I would probably prefer a callback api, rather than promises, we don't use so much promises in our core code.

Reply:
Yeah, SGTM. Found some hiccups while refactoring to callback, I'll go through them later on but so far so good 👍